oraclesrv/utils.py

```diff
@@ -0,0 +1,41 @@
+
+from flask import current_app
+import requests
+
+from oraclesrv.client import client
+
+def get_solr_data(reader, rows=5, sort='entry_date', cutoff_days=5, top_n_reads=10):
+    """
+
+    :param reader:
+    :param rows:
+    :param sort:
+    :param cutoff_days:
+    :param top_n_reads:
+    :return:
+    """
+    query = '(similar(topn({topn}, reader:{reader}, {sort} desc)) entdate:[NOW-{cutoff_days}DAYS TO *])'.format(
+                     topn=top_n_reads, reader=reader, sort=sort, cutoff_days=cutoff_days)
+    try:
+        response = client().get(
+            url=current_app.config['ORACLE_SERVICE_SOLRQUERY_URL'],
+            headers={'Authorization': 'Bearer ' + current_app.config['ORACLE_SERVICE_ADSWS_API_TOKEN']},
+            params={'fl': 'bibcode', 'rows': rows, 'q': query},
+        )
+        if response.status_code == 200:
```

Comment:
please add something like `response.raise_for_errors()` or also deal with the non-200 responses; otherwise it will be failing silently

Reply:
sorry. confused. so if solr returns other than 200, it is logged and returns error that did not get anything from solr. and if exception is logged. would you please give me more info. thank you
```
        if response.status_code == 200:
            from_solr = response.json()
            if from_solr.get('response'):
                num_docs = from_solr['response'].get('numFound', 0)
                if num_docs > 0:
                    current_app.logger.debug('Got {num_docs} records from solr.'.format(num_docs=num_docs))
                    result = []
                    for doc in from_solr['response']['docs']:
                        result.append(doc['bibcode'])
                    return result, query
        current_app.logger.error('Solr returned {response}.'.format(response=response))
        return None, query
    except requests.exceptions.RequestException as e:
        # catastrophic error. bail.
        current_app.logger.error('Solr exception. Terminated request.')
        current_app.logger.error(e)
        return None, query
```

Reply:
it is going to return None, it is not going to fail; so there is no way to distinguis empty response from errorred response (when status is 404, 503 etc)

but if you you response.raise_for_errors() it will be taken care of

Reply:
yes, it is going to return none and then here it is going to return that did not get anything from solr: https://github.com/golnazads/oracle_service/blob/master/oraclesrv/views.py#L112. that is not enough? I can include the error code from solr.
Cannot find any documentation on response.raise_for_errors(), can you point me to some code where we are using it. thank you.

Reply:
would you please let me know your thoughts on this approach. As you suggested for adsws to propagate the exception, did the same thing for solr, logging the error and then raise the exception. if solr did not return any results or return status code beside 200, along with the message no result from solr the status_code is logged.

Reply:
committed the code. thank you.

Reply:
I have to disagree, there are 3 ifs
```
if status
    if response....
       if numHits > 0
```

so 3 conditions that all have to collectively succeed; to retun 200

the rest will be mixed with all the errors; and that's what I find less than ideal. I also prefer to fail early (for non-200 repsonses; but that might be a personal preference -- it makes debugging easier)

another issue is logs: all of the requests, even if solr failed, will be recorded in logs with 200 (there will be ERROR entries in the logs, but no corresponding non-200 logs to be found; that makes debugging hard)